### ISSUE TITLE
RFC6120: Some error conditions can have a payload

### ISSUE THREAD

**Issue by horazont** [horazont/aioxmpp#226]
Most importantly, ``<gone/>``.

aioxmpp currently rejects the payload due to "unexpected text", since we use ChildEnum.

@sebastianriese IIRC, you need something like this for Jingle too, maybe we can work out a proper generic solution here.

**Comment by horazont** [comment]
So I have the following plan:

#### 1. Create Enum mixin

```python
class XSOEnum:
    def __init__(self, namespace, localname):
        super().__init__()
        self.xso_class = self._create_class((namespace, localname))

    def _create_name(self):
        return "".join(map(str.title, self.name.split("_"))),

    def _create_class(self):
        return aioxmpp.xso.model.XMLStreamClass(
            self._create_name(),
            (aioxmpp.xso.XSO,),
            {
                "TAG": self.value,
                "__qualname__": ...,
                "enum_member": self,
            },
        )

# example use

class ErrorCondition(XSOEnum):
    BAD_REQUEST = (namespaces.stanzas, "bad-request")
    ...
```

#### 2. Create helper function to create xso.Child descriptors

* Takes XSOEnum
* Creates xso.Child descriptor which accepts all the member.xso_class children

#### 3. Use simple wrapper properties to make access easier

```python
class Error(xso.XSO):
    condition_object = xso.child_xso_enum_helper(ErrorCondition)

    @condition
    def condition(self):
        if self.condition_object is None:
            return None
        return self.condition_object.enum_member

    @condition.setter
    def condition(self, value):
        if self.condition == value:
            return
        self.condition_object = value.xso_class()
```

#### Benefit

* Monkey patching allows to add payloads
* Access to the enum value is still simple

**Comment by sebastianriese** [comment]
Modulo the magic EnumMixin this is what I did for Jingle. Only disadvantage is, that we won't have a constructor that makes it easy to attach children (but that is certainly accetable).

**Comment by horazont** [comment]
You can just invoke the register_child function in a loop for that.

**Comment by horazont** [comment]
Fixed with #227 (which was merged, but github doesn’t get that).
